src/main/java/io/stargate/sgv2/jsonapi/service/schema/tables/ApiRegularIndex.java

```diff
@@ -112,59 +83,183 @@ public ApiRegularIndex create(
-                    map -> {
-                      map.put(
-                          "allColumns",
-                          errFmtApiColumnDef(tableSchemaObject.apiTableDef().allColumns()));
-                      map.put("unsupportedColumns", errFmt(apiColumnDef));
-                    }));
-          }
-        }
-        // nothing to update in the cqlOptions for these indexes
-      } else {
-        // text and ascii fields can have the text analysis options specified
-        Options.ASCII.putOrDefaultStringable(
-            indexOptions, optionsDesc == null ? null : optionsDesc.ascii());
-        Options.CASE_SENSITIVE.putOrDefaultStringable(
-            indexOptions, optionsDesc == null ? null : optionsDesc.caseSensitive());
-        Options.NORMALIZE.putOrDefaultStringable(
-            indexOptions, optionsDesc == null ? null : optionsDesc.normalize());
+      // Rule 4
+      String indexFunction =
+          apiColumnDef.type().typeName() == ApiTypeName.MAP
+              ? ApiIndexFunction.ENTRIES.cqlFunction
+              : ApiIndexFunction.VALUES.cqlFunction;
+      indexOptions.put("indexFunction", indexFunction);
+
+      return new ApiRegularIndex(
+          indexIdentifier, ApiIndexType.COLLECTION, columnIdentifier, indexOptions);
+    }
+
+    /**
+     * Populate the indexOptions map by pulling values from optionsDesc. indexOptions will be used
+     * to append the options in createIndex cql statement.
+     */
+    private void populateIndexOptionsMap(
+        Map<String, String> indexOptions,
+        RegularIndexDefinitionDesc.RegularIndexDescOptions optionsDesc) {
+      Options.ASCII.putOrDefaultStringable(
+          indexOptions, optionsDesc == null ? null : optionsDesc.ascii());
+      Options.CASE_SENSITIVE.putOrDefaultStringable(
+          indexOptions, optionsDesc == null ? null : optionsDesc.caseSensitive());
+      Options.NORMALIZE.putOrDefaultStringable(
+          indexOptions, optionsDesc == null ? null : optionsDesc.normalize());
+    }
+
+    /**
+     * Only text and ascii datatypes can be analyzed. <br>
+     * Only text and ascii in the collection datatype can be analyzed. It works for values(list),
+     * values(set), keys(map), values(map). Note, not for entries(map). <br>
+     * The method will error out as UNSUPPORTED_TEXT_ANALYSIS_FOR_DATA_TYPES if rules are violated.
+     */
+    private void validateAnalyzableDatatypes(
+        TableSchemaObject tableSchemaObject,
+        ApiColumnDef apiColumnDef,
+        RegularIndexDefinitionDesc.RegularIndexDescOptions optionsDesc,
+        RegularIndexDefinitionDesc indexDesc) {
+
+      if (optionsDesc == null) {
+        // no need to analyze if user does not specify any RegularIndexDescOptions
+        return;
+      }
+
+      // Primitive type
+      ApiTypeName analyzedColumnType = apiColumnDef.type().typeName();
+
+      // Map collection type
+      if (apiColumnDef.type() instanceof ApiMapType) {
```

Comment:
check the typeName

not sure why this errors for any map type? It is called from createApiIndexForCollection() 

we need to clarify what the rules for the options are. I thought none of the three text options would be allowed for set / list / map